# Hotel Booking Cancellation Prediction
  The main goal of this project is to understand hotel booking data and use machine learning to predict whether a customer will cancel their booking or not.

## 1.Project Introduction
    1.1 Problem Statement:
        Hotel bookings are sometimes cancelled by customers.These cancellations can cause problems for hotels, such as empty rooms ans loss of revenue. So, we want to understand the booking data and find the reasons behind cancellations.
    
    1.2 Business Objective:
        The objective is to help hotels identify bookings that are more likely to be cancelled. This can help them manage rooms and booking more effectively.
        
    1.3 Project Goal: 
        The goal of this project is to analyze hotel booking data and build a machine learning model that can predict whether a booking will be cancelled or not

## 2.Import Libraries

In [150]:
import pandas as pd

## 3.Load Dataset

In [151]:
# 3.1 Read Dataset
df = pd.read_csv("../data/hotel_bookings.csv")

In [152]:
#3.2 Display first 5 rows
df.head()

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


## 4.1 Dataset Understanding

In [153]:
#4.1 Dataset shape 
df.shape

(119390, 32)

In [154]:
#4.2 Column Names 
df.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date'],
      dtype='object')

In [155]:
#4.3 Data types
df.dtypes

hotel                              object
is_canceled                         int64
lead_time                           int64
arrival_date_year                   int64
arrival_date_month                 object
arrival_date_week_number            int64
arrival_date_day_of_month           int64
stays_in_weekend_nights             int64
stays_in_week_nights                int64
adults                              int64
children                          float64
babies                              int64
meal                               object
country                            object
market_segment                     object
distribution_channel               object
is_repeated_guest                   int64
previous_cancellations              int64
previous_bookings_not_canceled      int64
reserved_room_type                 object
assigned_room_type                 object
booking_changes                     int64
deposit_type                       object
agent                             

In [156]:
# 4.4 Missing values 
df.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

In [157]:
# 4.5 Duplicate values 
df.duplicated().sum()

np.int64(31994)

In [158]:
# 4.6 Target variable distribution
df["is_canceled"].value_counts()

is_canceled
0    75166
1    44224
Name: count, dtype: int64

In [159]:
#4.7 Statistical summary
df.describe()

,is_canceled,lead_time,arrival_date_year,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,children,babies,is_repeated_guest,previous_cancellations,previous_bookings_not_canceled,booking_changes,agent,company,days_in_waiting_list,adr,required_car_parking_spaces,total_of_special_requests
count,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119386.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,103050.000000,6797.000000,119390.000000,119390.000000,119390.000000,119390.000000
mean,0.370416,104.011416,2016.156554,27.165173,15.798241,0.927599,2.500302,1.856403,0.103890,0.007949,0.031912,0.087118,0.137097,0.221124,86.693382,189.266735,2.321149,101.831122,0.062518,0.571363
std,0.482918,106.863097,0.707476,13.605138,8.780829,0.998613,1.908286,0.579261,0.398561,0.097436,0.175767,0.844336,1.497437,0.652306,110.774548,131.655015,17.594721,50.535790,0.245291,0.792798
min,0.000000,0.000000,2015.000000,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,6.000000,0.000000,-6.380000,0.000000,0.000000
25%,0.000000,18.000000,2016.000000,16.000000,8.000000,0.000000,1.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,9.000000,62.000000,0.000000,69.290000,0.000000,0.000000
50%,0.000000,69.000000,2016.000000,28.000000,16.000000,1.000000,2.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,14.000000,179.000000,0.000000,94.575000,0.000000,0.000000
75%,1.000000,160.000000,2017.000000,38.000000,23.000000,2.000000,3.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,229.000000,270.000000,0.000000,126.000000,0.000000,1.000000
max,1.000000,737.000000,2017.000000,53.000000,31.000000,19.000000,50.000000,55.000000,10.000000,10.000000,1.000000,26.000000,72.000000,21.000000,535.000000,543.000000,391.000000,5400.000000,8.000000,5.000000


- The dataset contains 119,390 bookings.
- Around 37% of bookings were cancelled.
- The average booking was made around 104 days before arrival.
- Some columns contain missing values, such as children, country, agent, and company.
- Some coloumns have unusual values that need to be checked during the data-cleaning process.

# 5. Data Cleaning


### 5.1 Handling missing values
    found missing values in four coloumns, children, country, agent, company




In [160]:
# children:- children is a count, so it should be a whole number. The median is 0, so we will replace the 4 missing values with 0.
    
df['children'] = df['children'].fillna(0)

In [161]:
# country :- We use Unknown because we dont know the customers country, so we dont want to guess. 
df['country'] = df['country'].fillna("Unknown")

In [162]:
# agent :- We use 0 because a missing agent means no agent information was provided, and 0 represents that clearly. 
df['agent'] = df['agent'].fillna(0)

In [163]:
# company :- The company column was removed because most of its values were missing, making it unreliable and less useful for the prediction model. 
df.drop(columns=['company'], inplace=True)

### 5.2 Handling Duplicate Rows
    31,994 duplicate rows were found in the dataset. These duplicate records can give extra weight so the same booking and affect our analysis and model.

In [174]:
df = df.drop_duplicates()

### 5.3 Check Invalid Values
    One invalid negative adr value(-6.38) was found and removed because a room rate cannot be negative.

In [165]:
df = df[df['adr'] >= 0]

### 5.4 Check Data Consistency
    166 records had zero adults, children, and babies.Since a valid booking should have at least one guest, these records were removed as inconsistent data.

In [166]:
((df['adults'] ==0) & (df['children'] == 0) & (df['babies'] == 0)).sum()

np.int64(166)

In [167]:
invalid_rows = df[(df['adults'] == 0) & (df['children'] == 0) & (df['babies'] == 0)].index

In [168]:
df.drop(invalid_rows, inplace=True)

### 5.5 Data Leakage Coloumns
    The resiervation_status and reservation_status_date coloumns were removed because they contain information about the booking outcome. Keeping them could cause data leakage and make the model unrealistically accurate.

In [169]:
df[['reservation_status', 'reservation_status_date']].head()

,reservation_status,reservation_status_date
0,Check-Out,2015-07-01
1,Check-Out,2015-07-01
2,Check-Out,2015-07-02
3,Check-Out,2015-07-02
4,Check-Out,2015-07-03


In [170]:
df = df.drop(columns=['reservation_status', 'reservation_status_date'])